### PR TITLE
Make nodecontroller configure nodes' pod IP ranges

### PR DESCRIPTION
**Pull request by tkulczynski**
Moves podCIDRs assignment and routes creation in GCE from bash scripts to node controller. This is mainly to allow dynamic node set changes without complicated logic.

Tested: e2e tests pass, also if you manually resize up the MIG with nodes after initial cluster validation. I only tested resizing up because K8s won't handle nodes going away properly. Cluster validation verifies the number of nodes is as specified in the config (it's generally a good thing to have I believe, so didn't change it): I also made sure that validate-cluster passes after resizing the node group if you just manually set NUM_NODES to what you used for resize.

### CODE REVIEW

**Comment by tkulczynski** [comment]
Note: to some extent, this addresses #4610: restarting a node under the same name works now and you can add more nodes to a running cluster. Still, K8s might get confused if you delete nodes that have some pods running on them etc. - these are clearly next steps.

**Comment by bgrant0607** [comment]
This is a significant change. You also need to update docs, such as:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/networking.md
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/networking.md

and probably some of the getting-started guides.

**Comment by tkulczynski** [comment]
Brian, I'll go through those docs more thoroughly tomorrow and will try to look for other places that might be impacted, but at first glance there's not much to change - networking model stays mostly the same, just the implementation changes - some other component is responsible for assigning the same IP ranges to nodes. I agree it might be worth to add a word here and there, but I haven't yet seen a sentence in the docs that is made incorrect by my change.

**Comment by alex-mohr** [comment]
CJ has been working with Tomek on this plan, so switching assignee to him.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 1]
```diff
@@ -80,6 +80,49 @@ func NewNodeController(
...
+	for i := 0; i < num; i++ {
+		// TODO: Make the CIDRs configurable.
+		res.Insert(fmt.Sprintf("10.244.%v.0/24", i))
+	}
+	return res
+}
+
+// For each node from newNodes, finds its current spec in registeredNodes.
+// If it is not there, it gets a new valid CIDR assigned.
+func reconcilePodCIDRs(newNodes, registeredNodes *api.NodeList) *api.NodeList {
```
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: If you're getting a pointer to a struct passed in, you don't need to pass the pointer back to ensure the caller sees your changes.

File: pkg/cloudprovider/gce/gce.go [hunk 2]
```diff
@@ -421,6 +464,57 @@ func (gce *GCECloud) GetNodeResources(name string) (*api.NodeResources, error) {
 	}
 }
 
+func getMetadataValue(metadata *compute.Metadata, key string) string {
+	for _, item := range metadata.Items {
+		if item.Key == key {
+			return item.Value
+		}
+	}
+	return ""
```
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: "" is a valid value for a metadata item. Maybe consider a (string, bool) map-style interface for this function?

File: pkg/cloudprovider/gce/gce.go [hunk 3]
```diff
@@ -421,6 +464,57 @@ func (gce *GCECloud) GetNodeResources(name string) (*api.NodeResources, error) {
...
+}
+
+func (gce *GCECloud) Configure(name string, spec *api.NodeSpec) error {
+	instanceName := canonicalizeInstanceName(name)
+	instanceCall := gce.service.Instances.Get(gce.projectID, gce.zone, instanceName)
```
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: Take a look at the getMetadata function above. Each instance has a local (in the host OS) metadata server that it can curl directly. This avoids the network latency, and reduces quota usage for the GCE API.

File: pkg/cloudprovider/gce/gce.go [hunk 4]
```diff
@@ -421,6 +464,57 @@ func (gce *GCECloud) GetNodeResources(name string) (*api.NodeResources, error) {
...
+		DestRange:       spec.PodCIDR,
+		NextHopInstance: fmt.Sprintf("zones/%s/instances/%s", gce.zone, instanceName),
+		Network:         fmt.Sprintf("global/networks/%s", gce.networkName),
+		Priority:        1000,
+	})
+	insertOp, err := insertCall.Do()
+	if err != nil {
+		return err
+	}
+	return gce.waitForGlobalOp(insertOp)
```
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: Serializing metadata reads/writes and route inserts for each node is going to be horribly slow and wasteful of quota.

I'd suggest doing the metadata read/writes as a batch: Read metadata once, append all new CIDR keys, write back once.
Then kick off a goroutine to insert a route and wait on the op for each node.

This will require a little bit of restructuring up in SyncCloud.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 5]
```diff
@@ -80,6 +80,49 @@ func NewNodeController(
...
+			podCIDR, _ = availableCIDRs.PopAny()
+		}
+		newNodes.Items[i].Spec.PodCIDR = podCIDR
+	}
+	return newNodes
+}
+
+func (s *NodeController) configureNodeCIDR(node *api.Node) {
+	var err error
+	if instances, ok := s.cloud.Instances(); ok {
```
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: The go style guide would say that you should invert the logic here so that only the exceptional case is indented. I think that would also make the error flow easier to follow:

instances, ok := s.cloud.Instances()
if !ok {
	glog.Errorf("Error configuring node %s: CloudProvider does not support Instances()", node.Name)
	return
}
if err := instances.Configure(node.Name, &node.Spec); err != nil {
	glog.Errorf("Error configuring node %s: %s", node.Name, err)
	node.Spec.PodCIDR = ""
	return
}

File: pkg/cloudprovider/gce/gce.go (hunk 4 above)
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: I take back my comment about batch metadata (forgot this isn't project metadata). This still should be done in parallel.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: I don't get your comment - it seems to me that I'm doing exactly that?

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 5 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: I thought I won't duplicate the glog. But you're definitely right I've made it hard to read - changing.

File: pkg/cloudprovider/gce/gce.go (hunk 4 above)
Comment (tkulczynski) [inline pkg/cloudprovider/gce/gce.go]: Would you be happy with me replacing the synchronous loop over nodes in nodecontroller with scheduling a goroutine for each node that does all the work for the node and then waiting for all of routines? I think the cloudprovider serves as a synchronous client lib and I don't want to introduce new patterns here.

Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: I would be happier with that, although I don't have enough context with the architecture goals involved. I'll try to sync up with somebody today to get a better answer.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: Yes. I'm saying you don't need to:
func bar(*foo) *foo
where the pointer doesn't actually change can just be:
func bar(*foo)

I've seen this pattern elsewhere too. Is there some good reason for it that I am missing?

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 6]
```diff
@@ -108,7 +151,9 @@ func (s *NodeController) Run(period time.Duration, syncNodeList, syncNodeStatus
 			glog.Errorf("Error loading initial static nodes: %v", err)
 		}
 	}
+	nodes = s.DoChecks(nodes)
 	nodes, err = s.PopulateAddresses(nodes)
+	nodes = reconcilePodCIDRs(nodes, &api.NodeList{})
```
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: Is there a reason we need to do this both in Run() and in SyncCloud()? It would be nice if we could at least keep this CIDR-distribution stuff isolated in the in-the-cloud section.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (thockin) [inline pkg/cloudprovider/controller/nodecontroller.go]: The pattern is often used when you want to one-line a call.

```
var foo Foo
DoSomething(&foo)
OtherThing(&foo)
```

Vs

```
var foo Foo
OtherThing(DoSomething(&foo))
```

Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Oh, I misread your first comment. I don't need to do that, I just copied it over from some code I saw nearby, I'm not at all attached to it.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 6 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: The current implementation I'm proposing here makes an assumption that CIDR assignment is done at the point when you first see a Node (i.e. just before creating it in the apiserver). I thought it's a bit harder to have a logic that also looks for nodes that are in the apiserver but don't have the IP assigned (it's all a bit overcomplicated because we still need to support cloudproviders that do not have the possibility to reconfigure IP range on the fly...). I understand it might not be the best way to solve all this, happy to take any proposals.

File: pkg/cloudprovider/gce/gce.go (hunk 4 above)
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: Along with creating the routes in parallel, we also have to think about who tears down the routes (and how).

I don't really like the asymmetric route creation/deletion (i.e. nodecontroller.go creates routes, but util.sh cleans them up). But I'm not particularly excited about making the master responsible for deleting them either. Right now, GKE only cleans up what it created, so we'll have to look at how we're going to handle this.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: Thanks @thockin for the explanation. In that case, I think it's fine to keep the function signature as is, but we don't need the spurious assignments below at the call site.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 7]
```diff
@@ -108,7 +151,9 @@ func (s *NodeController) Run(period time.Duration, syncNodeList, syncNodeStatus
 			glog.Errorf("Error loading initial static nodes: %v", err)
 		}
 	}
+	nodes = s.DoChecks(nodes)
```
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: Why did DoChecks get added here?

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 6 above)
Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: It is awkward right now, because we assign CIDRs to nodes, but we only configure them with that CIDR if they are running in a cloud provider (and only in GCE right now). It is probably harmless, but it would be easier to follow if we only called reconcileCIDRs() when we are going to call configureCIDR().

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Done

File: pkg/cloudprovider/gce/gce.go (hunk 2 above)
Comment (tkulczynski) [inline pkg/cloudprovider/gce/gce.go]: Done

File: pkg/cloudprovider/gce/gce.go (hunk 4 above)
Comment (tkulczynski) [inline pkg/cloudprovider/gce/gce.go]: Added a symmetric Route teardown. It will now all work fine if somebody first deletes the nodes, waits a while and only then deletes the master. I'm not sure it's enough (I bet we want to keep some teardown code in bash anyway), but at least it's symmetric.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 5 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Done.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 6 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: Added the if also for reconciling.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 7 above)
Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: I honestly have no idea. Deleted.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 8]
```diff
@@ -80,6 +80,61 @@ func NewNodeController(
 	}
 }
 
+// Generates num pod CIDRs that could be assigned to nodes.
+func generateCIDRs(num int) util.StringSet {
+	res := util.NewStringSet()
+	for i := 0; i < num; i++ {
+		// TODO: Make the CIDRs configurable.
+		res.Insert(fmt.Sprintf("10.244.%v.0/24", i))
```
Comment (zmerlynn) [inline pkg/cloudprovider/controller/nodecontroller.go]: I have a diff sitting in my tree to make this configurable via env variable (something like KUBE_CLASS_B). Can you fix this sooner rather than later? It affects things like the ability to spin up multiple clusters in the same network, and networks are a slightly harder resource to come by.

(Yes, I realize it means some ugly salt config. :( )

Comment (tkulczynski) [inline pkg/cloudprovider/controller/nodecontroller.go]: I know it can be considered a step back, because I'm deleting similar code from the bash script that might (if somebody had set some non-standard env parameters) have been taking care of this and just leaving out a TODO here. Note though that the nodecontroller still accepts nodes configured by someone else, especially those that already have the ip-range metadata set.
Are you suggesting to resolve this TODO in this PR or just mentioning it's important in general?

**Comment by tkulczynski** [comment]
BTW, I've now verified that I can resize the node group multiple times both up and down and that the cluster is still operational (bu running validation after resizes, e.g. $ NUM_MINIONS=7 ./cluster/validate-cluster.sh) and that the routes are created and deleted alongside those changes. 
I don't feel too confident about removing routes deletion from from kube-down script, because we'd need to delete the nodes first, wait for the master to react and clean-up those and only then delete it, which seems much more complicated to me. I'm taking similar approach as monitoring services - creation is done by the master, but clean-up in the kube-down.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 8 above)
Comment (zmerlynn) [inline pkg/cloudprovider/controller/nodecontroller.go]: I'm saying that I might get around to checking that in before this PR even lands, so this could end up being a regression in configurability. It's only #4596 that stopped me from checking the rest of that script in (that commit is sitting in the same branch, for context).

Comment (zmerlynn) [inline pkg/cloudprovider/controller/nodecontroller.go]: (Yes, I opened that 6 weeks ago, but I was playing around with about 4 days ago after recent `.kubeconfig` stuff landed. It's still not in good shape for this usecase, unfortunately, but it's better.)

Comment (cjcullen) [inline pkg/cloudprovider/controller/nodecontroller.go]: Oooo yeah I missed this first time around. We need this range to be dynamic. I'll put together a PR that plumbs this through.

File: pkg/cloudprovider/gce/gce.go [hunk 9]
```diff
@@ -421,6 +464,67 @@ func (gce *GCECloud) GetNodeResources(name string) (*api.NodeResources, error) {
...
+			return item.Value, true
+		}
+	}
+	return "", false
+}
+
+func (gce *GCECloud) Configure(name string, spec *api.NodeSpec) error {
+	instanceName := canonicalizeInstanceName(name)
+	instanceCall := gce.service.Instances.Get(gce.projectID, gce.zone, instanceName)
+	instance, err := instanceCall.Do()
```
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: You could one-line this:
instance, err := gce.service.Instances.Get(gce.projectID, gce.zone, instanceName).Do()

**Comment by brendandburns** [comment]
Closing per pull-request policy, since inactive for > 14 days.  Please re-open when it is ready for review.

**Comment by tkulczynski** [comment]
I addressed the minor comment from cjcullen and rebased.

**Comment by jszczepkowski** [comment]
@cjcullen 
Can you please take a look? All comments are applied now, and this PR is vital for us to allow resizing of clusters.

File: pkg/cloudprovider/nodecontroller/nodecontroller.go [hunk 10]
```diff
@@ -136,6 +137,61 @@ func NewNodeController(
 	}
 }
 
+// Generates num pod CIDRs that could be assigned to nodes.
+func generateCIDRs(num int) util.StringSet {
+	res := util.NewStringSet()
+	for i := 0; i < num; i++ {
+		// TODO: Make the CIDRs configurable.
+		res.Insert(fmt.Sprintf("10.244.%v.0/24", i))
```
Comment (zmerlynn) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: This part needs to be addressed now that `KUBE_GCE_CLUSTER_CLASS_B` is in. (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/config-default.sh#L66)

**Comment by roberthbailey** [comment]
It might make sense for this to wait until #6949 goes in, which makes some significant changes to the nodecontroller code.

**Comment by jszczepkowski** [comment]
When will #6949 be ready? We need this PR soon to start testing node addition. I was hoping it will be even merged tomorrow. Resizing clusters is part of 1.0.

**Comment by roberthbailey** [comment]
@jszczepkowski Not for a few more days (need to fix some tests before re-pushing my changes). I certainly don't want to hold you up, but I worry that with nodes self registering we will need to entirely re-think the way that we distribute node CIDRs. I certainly won't object to this going in before my change once you address feedback from @cjcullen and @zmerlynn.

File: pkg/cloudprovider/nodecontroller/nodecontroller.go [hunk 11]
```diff
@@ -196,21 +255,30 @@ func (nc *NodeController) registerNodes(nodes *api.NodeList, retryCount int, ret
...
+		for i := range nodes.Items {
+			go func(node *api.Node) {
+				defer wg.Done()
+				if registered.Has(node.Name) {
+					return
+				}
+				if nc.isRunningCloudProvider() {
+					nc.configureNodeCIDR(node)
+				}
+				_, err := nc.kubeClient.Nodes().Create(node)
```
Comment (cjcullen) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: Do we know if kubeClient is thread-safe to be able to do this inside a goroutine?

File: pkg/cloudprovider/nodecontroller/nodecontroller.go (hunk 10 above)
Comment (cjcullen) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: This should also be extended to accommodate large clusters. KUBE_GCE_CLUSTER_CLASS_B implies it will only ever be a /16, but I think @satnam6502 has been trying to test larger clusters, and GKE uses a /14.

**Comment by cjcullen** [comment]
Could we put this functionality behind a flag in the NodeController? That would let us start to test node addition while also letting us think about a cleaner design post #6949. It would also prevent us from having to accommodate the change in GKE immediately (even though we are almost certainly going to have to eventually).

File: pkg/cloudprovider/nodecontroller/nodecontroller.go (hunk 11 above)
Comment (jszczepkowski) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: I think kubeClient *is* thread safe. For example: we are the same kubeClient from different threads in integration test.

File: pkg/cloudprovider/nodecontroller/nodecontroller.go (hunk 10 above)
Comment (jszczepkowski) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: I've added a flag for KUBE_GCE_CLUSTER_CLASS_B and made it configurable. Regarding range /14 for GKE: Tomek is working on GKE support, we will do it in a separate PR. (This PR is for GCE only).

**Comment by jszczepkowski** [comment]
@cjcullen
All comments are applied, PR is re-based and ready for review, PTAL.

Regarding extra flag: I don't know how to put this functionality behind a flag in NodeController, as we also change configuration scripts. So, if the functionality is turned off, GCE will just not work due to modifications in scripts. Details of integration with #6949 will be discussed on a conference today.

**Comment by tkulczynski** [comment]
Theoretically, we can postpone the changes in scripts if that's needed. I'm guessing the nodeController code shouldn't really need to be under a flag: if somebody else sets the node-ip-range metadata, the controller will not change it, that should be enough (although I don't feel strongly here and may not remember all the details, please verify my thinking).

**Comment by cjcullen** [comment]
@tkulczynski I think you are right functionally. What I want to avoid is a field that is always set, but only used by GCE/GKE. And currently, this flat breaks GKE :).

Even after we get GKE to handle the new form of deployment, we will continue to have providers that don't want the master to allocate CIDRs, so I think a "--allocate_node_cidrs" flag defaulted to false would be a good way to get this going.

**Comment by roberthbailey** [comment]
> I think a "--allocate_node_cidrs" flag 

nit: @thockin just switch all of our flags from under_score to dash-separated, so please keep that style when adding new ones.

File: pkg/cloudprovider/gce/gce.go [hunk 12]
```diff
@@ -154,6 +178,7 @@ func newGCECloud(config io.Reader) (*GCECloud, error) {
 		projectID:        projectID,
 		zone:             zone,
 		instanceID:       instanceID,
+		networkName:      networkName,
```
Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: Which component is calling this? We can't assume the nodes are on the same network as the master.

File: pkg/cloudprovider/gce/gce.go [hunk 13]
```diff
@@ -507,6 +551,66 @@ func (gce *GCECloud) GetNodeResources(name string) (*api.NodeResources, error) {
...
+		return err
+	}
+	if currentValue, ok := getMetadataValue(instance.Metadata, podCIDRMetadataKey); ok {
+		if currentValue == spec.PodCIDR {
+			// IP range already set to proper value.
+			return nil
+		}
+		return ErrMetadataConflict
+	}
+	instance.Metadata.Items = append(instance.Metadata.Items,
```
Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: What script/program/component is consuming this metadata?  A comment would be useful.

File: pkg/cloudprovider/gce/gce.go (hunk 12 above)
Comment (tkulczynski) [inline pkg/cloudprovider/gce/gce.go]: It's used for the Routes creation. And you're right: I did actively assume that - does the cluster even work when they are not? How would the master instance contact the node instances?
Anyway, if we do not want that assumption, we probably need to pass in the node network name to the master and into the gce client.

**Comment by tkulczynski** [comment]
CJ: if the Configure method on a cloud provider fails (which should be the case for cloud providers other than GCE currently), we do not set the ip-range field. I think this is close to best what we can do (it would be best to compute/read it from somewhere and set it anyway, but that might be hard and different for each provider).

**Comment by cjcullen** [comment]
I'll merge this tomorrow after 0.16.0 is cut. Then we can take 2 weeks to get this and GKE to agree on how to deploy routes before the next release :).

File: pkg/cloudprovider/gce/gce.go (hunk 12 above)
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: Agree with @bgrant0607, but I think this is fine for this PR. I'll plumb a node network name through the master in a separate PR.

Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: We're actively eliminating all cases where the master contacts the nodes for management purposes.

The only remaining cases should be the "proxy" functionality: proxy, redirect, port forwarding, exec, logs. That functionality will eventually be factored out into a separate component that could run in the same network as the kubelets, and the apiserver could proxy/redirect to it via a public IP.

Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: See also https://github.com/GoogleCloudPlatform/kubernetes/issues/3481#issuecomment-74937184

Comment (jszczepkowski) [inline pkg/cloudprovider/gce/gce.go]: I've added a TODO.

File: pkg/cloudprovider/nodecontroller/nodecontroller.go [hunk 14]
```diff
@@ -194,21 +259,30 @@ func (nc *NodeController) registerNodes(nodes *api.NodeList, retryCount int, ret
...
+				defer wg.Done()
+				if registered.Has(node.Name) {
+					return
+				}
+				if nc.isRunningCloudProvider() && nc.allocateNodeCIDRs {
+					nc.configureNodeCIDR(node)
+				}
+				_, err := nc.kubeClient.Nodes().Create(node)
+				if err == nil || apierrors.IsAlreadyExists(err) {
+					registered.Insert(node.Name)
```
Comment (jszczepkowski) [inline pkg/cloudprovider/nodecontroller/nodecontroller.go]: It seems there is a synchronization error here: ``registered'' set is not thread-safe.

**Comment by jszczepkowski** [comment]
I addressed all the comments (flag ``allocate_node_cidrs'' is added). However, there are still some issues with this PR:
1. e2e tests are failing,
2. there seem to be a synchronization error in https://github.com/GoogleCloudPlatform/kubernetes/pull/5246#discussion_r29340084.

I'm on holidays till next Tuesday, maybe @tkulczynski will try to take over this PR if he has time.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 15]
```diff
@@ -141,6 +143,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int64Var(&s.NodeMilliCPU, "node-milli-cpu", s.NodeMilliCPU, "The amount of MilliCPU provisioned on each node")
 	fs.Var(resource.NewQuantityFlagValue(&s.NodeMemory), "node-memory", "The amount of memory (in bytes) provisioned on each node")
 	fs.StringVar(&s.ClusterName, "cluster-name", s.ClusterName, "The instance prefix for the cluster")
+	fs.StringVar(&s.ClusterClassB, "cluster-class-b", "10.244", "Class B network address for Pods in cluster.")
```
Comment (roberthbailey) [inline cmd/kube-controller-manager/app/controllermanager.go]: We should make this a CIDR instead of a Class B. For GKE we use a /14 (which is larger than a Class B).

File: pkg/cloudprovider/gce/gce.go [hunk 16]
```diff
@@ -219,12 +245,12 @@ func (gce *GCECloud) targetPoolURL(name, region string) string {
 	return fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/regions/%s/targetPools/%s", gce.projectID, region, name)
 }
 
-func (gce *GCECloud) waitForRegionOp(op *compute.Operation, region string) error {
+func waitForOp(op *compute.Operation, getOperation func() (*compute.Operation, error)) error {
 	pollOp := op
 	for pollOp.Status != "DONE" {
 		var err error
 		time.Sleep(time.Second)
```
Comment (roberthbailey) [inline pkg/cloudprovider/gce/gce.go]: can you add a TODO here to add some backoff / jitter?

Comment (jszczepkowski) [inline pkg/cloudprovider/gce/gce.go]: Done.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 15 above)
Comment (jszczepkowski) [inline cmd/kube-controller-manager/app/controllermanager.go]: It should also work for /14. Please note that we use the third byte for identifying node and the last byte for identifying pods inside node. We use first two bytes for the whole network, so, if GKE uses /14 last two bits of the second bytes will be always set to 0.

If you insist on converting this to CIDR, I can do it in a separate PR, as this PR is already waiting much too long.

**Comment by roberthbailey** [comment]
Does this pass e2e tests yet?

**Comment by jszczepkowski** [comment]
Yes e2e are passing.

**Comment by cjcullen** [comment]
You need some sort of synchronization around the modifications of nodeMap in syncCloudNodes(). I'd recommend just creating a nodeMapLock for now.

**Comment by cjcullen** [comment]
If you want to rebase and address the synchronization issues. I can merge this tomorrow morning.

If you're tired of going back and forth, I can just merge #7799 :).

**Comment by roberthbailey** [comment]
I'm tired of going back and forth so I merged #7799. We can iterate on making that code better.

**Comment by roberthbailey** [comment]
Please keep an eye on Jenkins now that this is in. The first run after it was merged failed because the cluster infrastructure pods were not all running before we hit the timeout. I'm wondering if this change slows down cluster formation such that we need to adjust some timeouts.

**Comment by tkulczynski** [comment]
Thanks CJ, Jerzy, Robert and all the other's who helped with this PR! And sorry for not doing enough to have this merged in quickly.
